abdiff/core/run_ab_transforms.py

```diff
@@ -1 +1,203 @@
+    for input_file in input_files:
+        source = (input_file.split("/")[-1]).split("-")[0]
+        for docker_image, transformed_directory in run_configs:
+            container = DockerClient.containers.run(
+                docker_image,
+                command=generate_transform_command(source, input_file),
+                detach=True,
+                environment={
+                    "AWS_ACCESS_KEY_ID": CONFIG.AWS_ACCESS_KEY_ID,
+                    "AWS_SECRET_ACCESS_KEY": CONFIG.AWS_SECRET_ACCESS_KEY,
+                    "AWS_SESSION_TOKEN": CONFIG.AWS_SESSION_TOKEN,
+                },
+                labels={
+                    "docker_image": docker_image,
+                    "source": source,
+                    "input_file": input_file,
+                    "run": run_id,
+                },
+                volumes=[f"{transformed_directory}:/tmp"],
+            )
+            logger.info(
+                f"Container '{container.id}' (Docker image: {docker_image}) "
+                f"RUNNING transform for '{source}' input_file: {input_file}."
+            )
+            container_ids.append(container.id)
+
+    # wait for all container runs to complete (status = "exited")
+    # get logs from all Docker containers
+    get_logs(run_id, container_ids, log_file=f"{run_directory}/transformed/logs.txt")
+
+    # clone run data and add list of input files
+    run_data = read_run_json(run_directory)
+    run_data.update({"input_files": input_files})
+    update_or_create_run_json(run_directory, run_data)
+
+    elapsed_time = perf_counter() - start_time
+    logger.info(
+        "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
+    )
+    return get_transformed_files(run_directory)
+
+
+def get_transformed_files(run_directory: str) -> tuple[list[str], ...]:
+    """Get list of filepaths to transformed JSON files.
+
+    Args:
+        run_directory (str): Run directory.
+
+    Returns:
+        tuple[list[str]]: Tuple containing lists of paths to transformed
+            JSON files for each image, relative to 'run_directory'.
+    """
+    filepaths = []
+    for root, _, files in os.walk(Path(run_directory) / "transformed"):
+        sub_filepaths = []
+        if root.endswith("transformed"):
+            # skip 'transformed' root directory
+            continue
+        for filename in files:
+            relative_dir = root.replace(run_directory, "")
+            sub_filepaths.append(os.path.join(relative_dir, filename))
+        filepaths.append(sub_filepaths)
+    return tuple(filepaths)
+
+
+def get_logs(run_id: str, container_ids: list[str], log_file: str) -> str:
+    """Monitor Docker container runs and write logs to file as containers exit."""
+    logged_container_ids: set[str] = set()
+    logs = ""
+    while logged_container_ids != set(container_ids):
+        if not DockerClient.containers.list(all=True, filters={"label": f"run={run_id}"}):
```

Comment:
If we know the list of container ids we care about, why do we need to filter by the label `run`?   Would it be possible to just get _all_ containers, and then filter _them_ to only those in the passed container ids?  

And/or, looks like you can filter to the "ancestor", which is the image of the container?

[From docs](https://docker-py.readthedocs.io/en/stable/containers.html?highlight=def%20list#docker.models.containers.ContainerCollection.list):
```
- `ancestor` (str): Filter by container ancestor. Format of
    ``<image-name>[:tag]``, ``<image-id>``, or
    ``<image@digest>``.
```

I ask, as the `run_id` jumped out at me in the main function.  I think it _is_ the run ID for all intents and purposes, but it's kind of the first time we mentioned it as such.  Until now, we've just referred to a run by `run_directory`.  I don't think it's wrong, but just noting this is the first time we kind of elevate the timestamp to an "id" status.

Like.... maybe each run _should_ have an ID?  We could store this in the `run.json`?  Just some thoughts.

Reply:
Good points, I'm all in favor of minimizing the "IDs" we have to keep track of so if possible sticking with the `run_directory` to ID the run

Reply:
For the record, I might propose that we do create unique IDs for each run, like a UUID.  We can store this in the run JSON, and maybe we'll never need it, but it would be a way to absolutely unambiguously reference every run within a job.

Reply:
I agree that a run UUID would be helpful! While writing / testing the code above, I was executing a lot of Docker containers, and when I looked at Docker Desktop, I realized I had a long list of exited Docker containers and didn't have a way to connect the executed Docker container with a specific run. 🤔 This led to the decision to label the Docker containers with a "run ID", which -- as the code is currently written -- would translate to the lowest level of the run directory: the run timestamp (e.g., `2024-09-05_11-42-00`).

The first `if` condition may not be necessary, but I do think filtering by the "run ID" is a requirement as _all_ exited Docker containers (including containers from other runs) will be returned otherwise. The result will also be the same if we filter by ancestor (i.e., one docker image can be associated with many runs) Let me know what you think given this additional information!

Reply:
Trying to be succinct here, and summarize the discussion above: minting a run UUID makes sense for the reasons you outline @jonavellecuerdo.  When we start a run, we know we're going to start `n` containers, and we know we want to wait for _only_ those containers to finish.  Therefore, if we know each container will have a tag/label of the run UUID, that feels like an airtight way to filter down to only those containers as ones we're interested in.

But, I'd be remiss if I didn't ask a followup question that maybe got buried above.

[Right here](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/TIMX-341-run-ab-transforms/abdiff/core/run_ab_transforms.py#L52), when we kickoff a container, it feels like we have the _specific_ container in hand in code.  If we collect those in a list, couldn't we just loop through those containers and see when they're done?  Or maybe I'm thinking about that wrong, that perhaps they don't provide a `.done()`, `.finished()`, `.still_running` method or something?

If we use the actual container instances that we have in python memory, I suppose I wonder why we'd need to query the `docker_client` for containers, as we already have them.

Reply:
Ah! That is a good point. Instead of collecting container IDs, collect the Container objects themselves -- taking a more object-oriented approach! Let me ponder this and see what changes can be made from this!

Reply:
Mentioned in slack, but noting here as well: I think the collected `Container` objects do _not_ automatically update their status, health, etc.  But I believe you can call `Container.reload()` which will update things like `.status`.

Reply:
Resolving this comment for now! Will continue to explore if there are better ways to simplify the code for waiting + log retrieval.